Fabric.Authorization.API/scripts/AzureADUsers-Migration.ps1

```diff
@@ -0,0 +1,390 @@
+if ($azureAD.length -eq 0) {
+    # Do not show error when AzureAD is not installed, will install instead
+    $installed = Get-InstalledModule -Name AzureAD -ErrorAction "silentlycontinue"
+
+    if (($null -eq $installed) -or ($installed.Version.CompareTo($minVersion) -lt 0)) {
+        Write-Host "Installing AzureAD from Powershell Gallery"
+        Install-Module AzureAD -Scope CurrentUser -MinimumVersion $minVersion -Force
+        Import-Module AzureAD -Force
+    }
+}
+else {
+    Write-Host "Installing AzureAD at $($azureAD.FullName)"
+    Import-Module -Name $azureAD.FullName
+}
+
+# Need the User principal methods for talking to AD
+Import-Module -Name .\Install-Authorization-Utilities.psm1 -Force
+
+# Import Identity Install Utilities
+# Used for getting access tokens
+$identityInstallUtilities = ".\Install-Identity-Utilities.psm1"
+if (!(Test-Path $identityInstallUtilities -PathType Leaf)) {
+    Write-DosMessage -Level "Warning" -Message "Could not find identity install utilities. Manually downloading and installing"
+    Invoke-WebRequest -Uri https://raw.githubusercontent.com/HealthCatalyst/Fabric.Identity/master/Fabric.Identity.API/scripts/Install-Identity-Utilities.psm1 -Headers @{"Cache-Control" = "no-cache"} -OutFile $identityInstallUtilities
+}
+Import-Module -Name $identityInstallUtilities -Force
+
+# Import Fabric Install Utilities
+$fabricInstallUtilities = ".\Fabric-Install-Utilities.psm1"
+if (!(Test-Path $fabricInstallUtilities -PathType Leaf)) {
+    Write-DosMessage -Level "Warning" -Message "Could not find fabric install utilities. Manually downloading and installing"
+    Invoke-WebRequest -Uri https://raw.githubusercontent.com/HealthCatalyst/InstallScripts/master/common/Fabric-Install-Utilities.psm1 -Headers @{"Cache-Control" = "no-cache"} -OutFile $fabricInstallUtilities
+}
+Import-Module -Name $fabricInstallUtilities -Force
+$credentials = @{}
+
+function Get-AuthUsers
+{	
+    param(
+        [Parameter(Mandatory=$true)]
+        [string] $connectionString
+    )
+    $connection = New-Object System.Data.SqlClient.SqlConnection($connectionString)	
+    $sql = "SELECT u.Id, u.SubjectId, u.IdentityProvider, u.IdentityProviderUserPrincipalName, g.GroupId, r.RoleId, p.PermissionId, p.PermissionAction, CAST('' AS NVARCHAR(100)) AS 'sourceAnchor', CAST('' AS NVARCHAR(100)) AS 'objectId'
+            FROM [dbo].[Users] u	
+            LEFT OUTER JOIN [dbo].[GroupUsers] g	
+                ON u.SubjectId = g.SubjectId AND u.IdentityProvider = g.IdentityProvider
+		    LEFT OUTER JOIN [dbo].[RoleUsers] r
+                ON u.SubjectId = r.SubjectId AND u.IdentityProvider = r.IdentityProvider
+            LEFT OUTER JOIN [dbo].[UserPermissions] p	
+                ON u.SubjectId = p.SubjectId AND u.IdentityProvider = p.IdentityProvider
+            WHERE u.IsDeleted = 0
+				AND u.IdentityProvider = 'Windows'
+				AND u.IdentityProviderUserPrincipalName IS NOT NULL";	
```

Comment:
Is this guaranteed to be populated when this script is run?

Reply:
If it meets the users meet the where constraint.

Reply:
I was asking if `IdentityProviderUserPrincipalName` guaranteed to be `NOT NULL`?

Reply:
I thought that this value was getting populated, but it doesn't look like Name or IdentityProviderUserPrincipalName are always there. For instance, when we install Auth and it adds the user, the Name and IdentityProviderUserPrincipalName aren't added, but if a user is added in Access Control, I believe it is added to IdentityProviderUserPrincipalName each time. So I may have to use SubjectId instead. What are your findings?

Reply:
I didn't think it was always non-null, so I think you'll need to use another field.

Reply:
I will just use SubjectId since it has to have a value. IdentityProviderUserPrincipalName was a column added later, and we avoided having to do a migration script by adding it as a nullable. I think the intention though was to always populate it.